submariner-operator/issue-149-node-placement.md

```diff
@@ -0,0 +1,96 @@
+# Node Placement
+
+[Add support for nodeSelector and toleration scheduling for submariner pods](https://github.com/submariner-io/enhancements/issues/149)
+
+## Summary
+
+Submariner deploys many different agents as part of its core deployment. Gateway, Globalnet Agent and
+Metrics Proxy are deployed on designated gateway nodes, while Route Aents are deployed on all nodes. There is no way
+for users to control or limit where rest of the agent pods are deployed. Some customers want to run these agents on
+specific nodes for billing or networking reasons. Users can't add tolerations directly to these agents coz they'll be
+overwritten by the operator. So, users require an easily configurable way to specify such configurations.
+
+## Proposal
+
+This enhancement proposes adding a new field to `Submariner` and `ServiceDiscovery` for users to configure node
+scheduling. This configuration will then be applied to pods of following submariner agents:
+
+* submariner-operator
+* submariner-lighthouse-agent
+* submariner-lighthouse-coredns
+
+Following agents will be excluded because these are run on Gateway nodes. Users already control which nodes get labeled
+as gateways and ideally we don't want to run anything else on these nodes.
+
+* submariner-gateway
+* submariner-globalnet
+* submariner-metrics-proxy
+
+## Design Details
+
+Add a new `NodePlacement` field to `Submariner` and `ServiceDiscovery` specs.
```

Comment:
So on official [K8s PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core) these fields are directly under it, I wonder if we don't want the same here?

AFAICT we can still bundle them up in a single `struct` to avoid repetitiveness, and use the as an embedded struct.

Reply:
> So on official [K8s PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core) these fields are directly under it, I wonder if we don't want the same here?
> 
I didn't get what you meant by "under it", could you clarify?

Reply:
Aha, got it. Since we're adding these fields to a generic Submariner spec, so `NodePlacement` gives a better idea about what these two fields are being used for. Also coz that is what the similar change I used for reference in openshift was using. But I do see your point, Node Selector and toleration imply used for placement of pods on specific nodes. Am okay to go with it [simpler and less nesting] unless anyone has other ideas.